File: swift/heat/heat_3.swift
//  Copyright (c) 2022 AUTHORS
//
//  SPDX-License-Identifier: BSL-1.0
//  Distributed under the Boost Software License, Version 1.0. (See accompanying
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
import Foundation

let biSema = DispatchSemaphore(value: 1)

let start = Date()

let C_ARGV = CommandLine.arguments

let nx = Int(C_ARGV[3]) ?? -1  // number of nodes
let k = 0.4  // heat transfer coefficient
let dt = 1.0  // time step
let dx = 1.0  // grid spacing
let nt = Int(C_ARGV[2]) ?? -1  // number of time steps
let threads = Int(C_ARGV[1]) ?? -1  // numnber of threads

struct Worker {

  let space: [UnsafeMutableBufferPointer<Double>]
  var num: Int
  var lo: Int
  var hi: Int
 
  init(){

    num = -1
    lo = -1
    hi = -1
    space = []

  }

  init(_ p_num: Int, _ tx: Int) {

    num = p_num
    lo = tx * num
    hi = (tx + 1) * (num)

    space = [
      UnsafeMutableBufferPointer<Double>.allocate(capacity: num + 2),
      UnsafeMutableBufferPointer<Double>.allocate(capacity: num + 2),
    ]

    space[0][0] = Double(hi)

    for i in 1...(num) {
      space[0][i] = Double(lo + i)
    }
    space[0][num + 1] = Double(lo)

  }

  func update(_ t: Int) {

    let r = (k * dt / (dx * dx))

    let dst = space[(t + 1) % 2]
    let src = space[t % 2]

    for i in 1...(num) {

      dst[i] =
        (src[i]
          + r
          * (src[i - 1] - 2 * src[i] + src[i + 1]))
    }

  }

  func send_left(_ t: Int) -> Double {

    return space[t % 2][1]

  }

  func send_right(_ t: Int) -> Double {

    return space[t % 2][num]

  }

  func send_ghost(_ left: Worker, _ right: Worker, _ t: Int) {

    left.receiv_right(t, send_left(t))
    right.receiv_left(t, send_right(t))

  }

  func receiv_ghost(_ left: Worker, _ right: Worker, _ t: Int) {

    space[t % 2][0] = left.send_right(t)
    space[t % 2][num + 1] = right.send_left(t)

  }

  func receiv_right(_ t: Int, _ value: Double) {

    space[(t + 1) % 2][num + 1] = value

  }

  func receiv_left(_ t: Int, _ value: Double) {

    space[(t + 1) % 2][0] = value

  }

}

let workerPool  = UnsafeMutableBufferPointer<Worker>.allocate(capacity: threads) 
let length = Int(nx / threads)

for t in 0...(threads - 1) {

  workerPool[t] = Worker(length, t)

}

for t in 0...(nt - 1) {

  await withTaskGroup(
    of: Void.self, returning: Void.self,
    body: { group in

      for p in 0...(threads - 1) {

        group.addTask {

          if threads > 1 {
            if p == 0 {
              workerPool[p].receiv_ghost(workerPool[threads - 1], workerPool[1], t)
            }

            else if p == threads - 1 {

               workerPool[p].receiv_ghost(workerPool[p - 1], workerPool[0], t)

            }

            else {

               workerPool[p].receiv_ghost(workerPool[p - 1], workerPool[p + 1], t)
            }
          }

          workerPool[p].update(t)

          if threads > 1 {
            if p == 0 {
               workerPool[p].send_ghost(workerPool[threads - 1], workerPool[1], t)
            }

            else if p == threads - 1 {

              workerPool[p].send_ghost(workerPool[p - 1], workerPool[0], t)

            }

            else {

               workerPool[p].send_ghost(workerPool[p - 1], workerPool[p + 1], t)
            }
          } else {

            workerPool[0].space[(t + 1) % 2][0] = workerPool[0].space[(t + 1) % 2][length]
             workerPool[0].space[(t + 1) % 2][length + 1] = workerPool[0].space[(t + 1) % 2][1]

          }

        }

      }

      for await _ in group {

      }

    })

}

print("swift,\(nx),\(nt),\(threads),\(dt),\(dx),\(-start.timeIntervalSinceNow)")

File: python/heat3.py
#  Copyright (c) 2023 AUTHORS
#
#  SPDX-License-Identifier: BSL-1.0
#  Distributed under the Boost Software License, Version 1.0. (See accompanying
#  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
from typing import Optional, Tuple, List, TypeVar, Generic
from threading import Thread, Lock, Condition
#from queue import Queue
import numpy as np
import sys
import time
import os

T = TypeVar('T')

class Queue(Generic[T]):
    def __init__(self):
        self.lock = Lock()
        self.items : List[T] = []
    def put_nowait(self, item: T)->None:
        with self.lock:
            self.items += [item]
    def get(self)->T:
        while True:
            with self.lock:
                if len(self.items)>0:
                    item = self.items[-1]
                    self.items = self.items[:-1]
                    return item
        time.sleep(1e-13)

use_hw_counters : bool = sys.argv[4] == "1"

check_correctness = False

if use_hw_counters:
    from pypapi import events, papi_high as high

ghosts = 1
nx = int(sys.argv[3])        # number of nodes
k = 0.4                      # heat transfer coefficient
dt = 1.                      # time step
dx = 1.                      # grid spacing
nt = int(sys.argv[2])        # number of time steps
threads = int(sys.argv[1])   # numnber of threads

class Worker(Thread):
    def __init__(self,num:int,tx:int)->None:
        Thread.__init__(self)
        self.num : int = num
        self.lo : int = tx*num
        self.hi : int = tx*(num+1)
        if self.hi > nx:
            self.hi = nx
        self.lo -= ghosts
        self.hi += ghosts
        self.right : Queue[float] = Queue()
        self.left : Queue[float] = Queue()
        self.sz = self.hi - self.lo
        assert self.sz > 0
        off = 1
        self.data = np.linspace(self.lo+off, self.hi-1+off, self.hi - self.lo)
        self.data2 = np.zeros((self.sz,))
        assert self.data.shape == self.data2.shape
        self.leftThread  : 'Worker'
        self.rightThread : 'Worker'

    def recv_ghosts(self)->None:
        self.data[0] = self.left.get()
        self.data[-1] = self.right.get()

    def update(self)->None:
        self.recv_ghosts()

        self.data2[1:-1] = self.data[1:-1] + (k * dt / (dx * dx)) * (self.data[2:] - 2*self.data[1:-1] + self.data[:-2])
        self.data, self.data2 = self.data2, self.data

        self.send_ghosts()

    def send_ghosts(self)->None:
        self.leftThread.right.put_nowait(self.data[1])
        self.rightThread.left.put_nowait(self.data[-2])

    def run(self)->None:
        self.send_ghosts()
        for n in range(nt):
            self.update()
        self.recv_ghosts()

def construct_grid(th:List[Worker]) -> np.ndarray:
    total = np.zeros((nx,))
    for t in th:
        total[t.lo + ghosts:t.hi - ghosts] = t.data[ghosts:-ghosts]

    print("Stats:",np.min(total),np.average(total),np.max(total))
    return total
    
def main(nthreads : int)->Tuple[float,float,np.ndarray]:
    th = []
    tx = (2*ghosts+nx)//nthreads
    assert tx > 0
    for num in range(nthreads):
        th += [Worker(num,tx)]
    for i in range(nthreads):
        th[i].rightThread = th[(i+1) % nthreads]
        th[(i+1)%nthreads].leftThread = th[i]

    if use_hw_counters:
        high.start_counters([events.PAPI_FP_OPS,])

    t1 = time.time()
    for t in th:
        t.start()
    for t in th:
        t.join()
    t2 = time.time()

    hw : int
    if use_hw_counters:
        hw = high.stop_counters()
    else:
        hw = 0

    total = construct_grid(th)
    avg0 = (nx+1)/2
    avgN = np.average(total)

    if nx < 20:
        print("grid:",total)
    print("time for ",nthreads,": ",t2-t1,sep="")
    err = np.abs(avgN - avg0)
    if err > 1e-13:
        print(f"{avgN} != {avg0}, err={err}")
    return t2-t1, hw, total

if check_correctness:
    tdiff, hw, data1 = main(1)
tdiff, hw, dataN = main(threads)
if check_correctness:
    print("diff check:",np.max(np.abs(dataN - data1)))
fn = 'perfdata.csv'
if not os.path.exists(fn):
    with open(fn,"w") as fd:
        print('lang,nx,nt,threads,dt,dx,total time,flops',file=fd)
with open("perfdata.csv","a") as fd:
    print(",".join(
        [str(x) for x in ['heat2', nx, nt, threads, dx, dt, tdiff, hw]]
    ),file=fd)
